ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp

```diff
@@ -529,6 +539,151 @@ bool OGRDXFWriterDS::TransferUpdateHeader(VSILFILE *fpOut)
             }
         }
 
+        // Patch INSUNITS
+        if (nCode == 9 && EQUAL(szLineBuf, "$INSUNITS") &&
+            m_osINSUNITS != "HEADER_VALUE")
+        {
+            if (!WriteValue(fpOut, nCode, szLineBuf))
+                return false;
+            nCode = oHeaderDS.ReadValue(szLineBuf, sizeof(szLineBuf));
+            if (nCode == 70)
+            {
+                int nVal = -1;
+                if (m_osINSUNITS == "FROM_CRS")
+                {
+                    if (m_oSRS.IsEmpty())
+                    {
+                        CPLError(CE_Warning, CPLE_AppDefined,
+                                 "Layer has no CRS. Using default value of "
+                                 "INSUNIT from template header file");
+                    }
+                    else if (m_oSRS.IsProjected())
+                    {
+                        const char *pszUnits = nullptr;
+                        const double dfUnits = m_oSRS.GetLinearUnits(&pszUnits);
+                        if (std::fabs(dfUnits - 1) <= 1e-10)
+                        {
+                            nVal = 6;
+                        }
+                        else if (std::fabs(dfUnits -
+                                           CPLAtof(SRS_UL_FOOT_CONV)) <= 1e-10)
+                        {
+                            nVal = 2;
+                        }
+                        else if (std::fabs(dfUnits -
+                                           CPLAtof(SRS_UL_US_FOOT_CONV)) <=
+                                 1e-10)
+                        {
+                            nVal = 21;
+                        }
+                        else
+                        {
+                            CPLError(CE_Warning, CPLE_AppDefined,
+                                     "Could not translate CRS unit %s to "
+                                     "INSUNIT. Using default value from "
+                                     "template header file",
+                                     pszUnits);
+                        }
+                    }
+                    else
+                    {
+                        CPLError(CE_Warning, CPLE_AppDefined,
+                                 "Layer CRS is not a projected CRS. "
+                                 "Using default value of INSUNIT from "
+                                 "template header file");
+                    }
+                }
+                else
+                {
+                    static const struct
+                    {
+                        const char *pszValue;
+                        int nValue;
+                    } INSUNITSMap[] = {
+                        {"UNITLESS", 0},
+                        {"INCHES", 1},
+                        {"FEET", 2},
+                        {"MILLIMETERS", 4},
+                        {"CENTIMETERS", 5},
+                        {"METERS", 6},
```

Comment:
Maybe we should be extra tolerant and also accept the UK spelling METRES.

Reply:
The declaration of the option uses an enumeration, so if users misspell they'll get an error. I'm not too entusiastic about tolerating all possible spelling variations (UK vs US, plural vs singular). The ones here are the ones of the DXF documentation